sw/device/tests/rstmgr_sw_rst_ctrl_test.c

```diff
@@ -0,0 +1,353 @@
+#include "sw/device/lib/dif/dif_i2c.h"
+#include "sw/device/lib/dif/dif_rstmgr.h"
+#include "sw/device/lib/dif/dif_spi_device.h"
+#include "sw/device/lib/dif/dif_spi_host.h"
+#include "sw/device/lib/dif/dif_usbdev.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_framework/ottf.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "i2c_regs.h"
+#include "spi_device_regs.h"
+#include "spi_host_regs.h"
+#include "usbdev_regs.h"
+
+/**
+ *  RSTMGR SW_RST_CTRL Test
+ *
+ *  This test checks RSTMGR.SW_RST_CTRL_N[index] with peripheral devices.
+ *  There are 8 RSTMGR.SW_RST_CTRL_N registers and each register
+ *  controls peripheral reset as follows:
+ *
+ * // index | device     |  test register |  reset value |  prgm value |
+ * // ------------------------------------------------------------------
+ * // 0     | SPI_DEVICE |  CFG           |  0x7f00      |  0x3f0c
+ * // 1     | SPI_HOST0  |  CONFIGOPTS    |  0x0         |  0x3210000
+ * // 2     | SPI_HOST1  |  CONFIGOPTS    |  0x0         |  0x6540000
+ * // 3     | USB        |  EP_OUT_ENABLE |  0x0         |  0xc3
+ * // 4     | USBIF      |  RXENABLE_OUT  |  0x0         |  0x695
+ * // 5     | I2C0       |  TIMING0       |  0x0         |  0x8b00cfe
+ * // 6     | I2C1       |  TIMING1       |  0x0         |  0x114010d8
+ * // 7     | I2C2       |  TIMING2       |  0x0         |  0x19ec1595
+ *
+ * 'test register' is a rw type register under each peripheral device.
+ * During the test, these registers are programmed with arbitrary values. ('prgm
+ * value') when sw resets are asserted, those values are updated to reset value.
+ * In each round, the rstmgr asserts reset to each peripheral device and
+ * each test register value is compared with the expected reset value.
+ */
+
+// peripheral device enum
+typedef enum node_type {
+  RSTMGR = 0,
+  SPI_DEVICE,
+  SPI_HOST0,
+  SPI_HOST1,
+  USB,
+  USBIF,
+  I2C0,
+  I2C1,
+  I2C2,
+  NUM_NODES
+} node_type_t;
+
+typedef struct node node_t;
+typedef void init_f(node_t *);
+typedef void config_f(node_t *);
+
+// peripheral device template
+// name : device name
+// base : base address
+// offset : offset to the test register
+// init_f : function ptr to each peripheral init
+// config_f : function ptr to each peripheral program
+// reset_val : reset value for comparison
+// prgm_val : expected value for test register after configuration
+//            function is called.
+struct node {
+  const char *name;
+  uintptr_t base;
+  ptrdiff_t offset;
+  init_f *init;
+  config_f *config;
+  uint32_t reset_val;
+  uint32_t prgm_val;
+};
+
+// handler declaration
+static dif_rstmgr_t rstmgr;
+static dif_spi_device_t spi_dev;
+static dif_spi_host_t spi_host0, spi_host1;
+static dif_usbdev_t usbdev;
+static dif_i2c_t i2c0, i2c1, i2c2;
+
+// test register read from a given peripheral device
+uint32_t read_test_reg(node_t *node) {
```

Comment:
static?

Reply:
removed

Reply:
Sorry what I meant was, this function can be made static, i.e., 
```c
static uint32_t read_test_reg(node_t *node) {
```

This applies throughout. A a general rule of thumb, any helper function defined in this test can be marked `static`.